# Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

Таблица books

Содержит данные о книгах:

book_id — идентификатор книги;

author_id — идентификатор автора;

title — название книги;

num_pages — количество страниц;

publication_date — дата публикации книги;

publisher_id — идентификатор издателя.

Таблица authors Содержит данные об авторах:

author_id — идентификатор автора;

author — имя автора.

Таблица publishers Содержит данные об издательствах:

publisher_id — идентификатор издательства;

publisher — название издательства;

Таблица ratings Содержит данные о пользовательских оценках книг:

rating_id — идентификатор оценки;

book_id — идентификатор книги;

username — имя пользователя, оставившего оценку;

rating — оценка книги.

Таблица reviews Содержит данные о пользовательских обзорах:

review_id — идентификатор обзора;

book_id — идентификатор книги;

username — имя автора обзора;

text — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [2]:
# Загружаем таблицу books
query = '''
            SELECT * FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.name = 'книги'

In [3]:
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
# Загружаем таблицу authors
query = '''
            SELECT * FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.name = 'имя автора'


In [5]:
       
authors

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [6]:
# Загружаем таблицу publishers
query = '''
            SELECT * FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.name = 'издательство'

In [7]:
publishers

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [8]:
# Загружаем таблицу ratings
query = '''
            SELECT * FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.name = 'Рейтинг'

In [9]:
ratings

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [10]:
# Загружаем таблицу reviews
query = '''
            SELECT * FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.name = 'Обзор книг'

In [11]:
reviews

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

# Изучение данных

In [12]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [13]:
#поиск дубликатов
books.duplicated().sum()

0

In [14]:
#категоризация
books.describe()

book_id    author_id   num_pages  publisher_id
count  1000.000000  1000.000000  1000.00000   1000.000000
mean    500.500000   320.417000   389.11100    171.270000
std     288.819436   181.620172   229.39014     99.082685
min       1.000000     1.000000    14.00000      1.000000
25%     250.750000   162.750000   249.00000     83.000000
50%     500.500000   316.500000   352.00000    177.500000
75%     750.250000   481.000000   453.00000    258.000000
max    1000.000000   636.000000  2690.00000    340.000000

In [15]:
#Поиск пропущенных значений
books.isnull().mean().reset_index()

index    0
0           book_id  0.0
1         author_id  0.0
2             title  0.0
3         num_pages  0.0
4  publication_date  0.0
5      publisher_id  0.0

In [16]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [17]:
#категоризация
authors.describe()

author_id
count  636.000000
mean   318.500000
std    183.741666
min      1.000000
25%    159.750000
50%    318.500000
75%    477.250000
max    636.000000

In [18]:
#поиск дубликатов
authors.duplicated().sum()

0

In [19]:
#Поиск пропущенных значений
authors.isnull().mean().reset_index()

index    0
0  author_id  0.0
1     author  0.0

In [20]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [21]:
#категоризация
publishers.describe()

publisher_id
count    340.000000
mean     170.500000
std       98.293777
min        1.000000
25%       85.750000
50%      170.500000
75%      255.250000
max      340.000000

In [22]:
#поиск дубликатов
publishers.duplicated().sum()

0

In [23]:
#Поиск пропущенных значений
publishers.isnull().mean().reset_index()

index    0
0  publisher_id  0.0
1     publisher  0.0

In [24]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [25]:
#категоризация
ratings.describe()

rating_id      book_id       rating
count  6456.000000  6456.000000  6456.000000
mean   3228.500000   510.574195     3.928284
std    1863.831001   284.141636     0.943303
min       1.000000     1.000000     1.000000
25%    1614.750000   291.000000     3.000000
50%    3228.500000   506.000000     4.000000
75%    4842.250000   750.000000     5.000000
max    6456.000000  1000.000000     5.000000

In [26]:
#поиск дубликатов
ratings.duplicated().sum()

0

In [27]:
#Поиск пропущенных значений
ratings.isnull().mean().reset_index()

index    0
0  rating_id  0.0
1    book_id  0.0
2   username  0.0
3     rating  0.0

In [28]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [29]:
#категоризация
reviews.describe()

review_id      book_id
count  2793.000000  2793.000000
mean   1397.000000   504.693161
std     806.413976   288.472931
min       1.000000     1.000000
25%     699.000000   259.000000
50%    1397.000000   505.000000
75%    2095.000000   753.000000
max    2793.000000  1000.000000

In [30]:
#поиск дубликатов
reviews.duplicated().sum()

0

In [31]:
#Поиск пропущенных значений
reviews.isnull().mean().reset_index()

index    0
0  review_id  0.0
1    book_id  0.0
2   username  0.0
3       text  0.0

# Подготовка данных

In [32]:
# изменим тип даты на соответствующий
books['publication_date'] = pd.to_datetime(books['publication_date'])

# Задачи

In [33]:
#Создадим функцию для запросов к данным:

def select(sql):
    return pd.read_sql(sql, con=engine)

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [34]:
books_count = '''
                SELECT 
                COUNT(title) AS books_count
                FROM
                    books
                WHERE publication_date >= '2000-01-01';
             ''' 
select(books_count)

books_count
0          821

После 1 января 2000 г. вышла 821 книга

<div class="alert-warning"> 
<b>Комментарий ревьюера 💡</b> 

Тут еще можно трактовать по-иному: буквально после 1 января 2000 года. Можно вывести оба варианта в идеале, но на вкус и цвет :)

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [35]:
query_count_avg = '''                              
                    SELECT
                        books.title AS book_title,
                        books.book_id AS book_id,
                        AVG(ratings.rating) AS avg_rating,
                        COUNT(DISTINCT reviews.review_id) AS count_reviews 

                    FROM
                        books

                    INNER JOIN ratings ON ratings.book_id = books.book_id
                    LEFT JOIN reviews ON reviews.book_id = books.book_id
                    

                    GROUP BY
                        books.title,
                        books.book_id

                    ORDER BY
                        count_reviews DESC;
                ''' 
select(query_count_avg).round(3)#.head(10)

book_title  book_id  avg_rating  \
0                              Twilight (Twilight  #1)      948       3.662   
1                                  Water for Elephants      963       3.977   
2                                     The Glass Castle      734       4.207   
3    Harry Potter and the Prisoner of Azkaban (Harr...      302       4.415   
4    The Curious Incident of the Dog in the Night-Time      695       4.081   
..                                                 ...      ...         ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel       83       3.667   
996                            The Natural Way to Draw      808       3.000   
997   The Cat in the Hat and Other Dr. Seuss Favorites      672       5.000   
998                          Essential Tales and Poems      221       4.000   
999  Disney's Beauty and the Beast (A Little Golden...      191       4.000   

     count_reviews  
0                7  
1                6  
2                6  
3                6  
4                6  
..             ...  
995              0  
996              0  
997              0  
998              0  
999              0  

[1000 rows x 4 columns]

Посчитали для каждой книги количество обзоров и среднюю оценку

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [36]:
publisher = '''
                     SELECT 
                         publishers.publisher_id, 
                         publishers.publisher, 
                         COUNT(books.book_id) as count_books
                     FROM 
                         publishers
                     JOIN books ON publishers.publisher_id = books.publisher_id
                     WHERE 
                         books.num_pages > 50
                     GROUP BY publishers.publisher_id
                     ORDER BY count_books DESC
                     LIMIT 5;
                 ''' 
select(publisher)

publisher_id                 publisher  count_books
0           212             Penguin Books           42
1           309                   Vintage           31
2           116  Grand Central Publishing           25
3           217          Penguin Classics           24
4            35                    Bantam           19

Издательство Penguin Books выпустило больше всего книг

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [37]:
rating_avg = '''
                        SELECT
                           authors.author_id AS id,
                           authors.author AS author,
                           AVG(ratings.rating) AS avg_rating,
                           COUNT(ratings.rating) AS count_rating
                        FROM
                           authors
                        INNER JOIN books ON books.author_id = authors.author_id
                        INNER JOIN ratings ON ratings.book_id = books.book_id
                        GROUP BY
                           id,
                           author
                        HAVING 
                           COUNT(ratings.rating) >= 50 
                           
                        ORDER BY
                           count_rating DESC
                        LIMIT 5
                        ;
                '''
select(rating_avg).round(3)

id                      author  avg_rating  count_rating
0  236  J.K. Rowling/Mary GrandPré       4.288           312
1  240              J.R.R. Tolkien       4.241           166
2  554             Stephenie Meyer       3.662           160
3  106                   Dan Brown       3.741           143
4  445             Nicholas Sparks       3.883           111

Автор с самой высокой средней оценкой книг J.K. Rowling/Mary GrandPré

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [40]:
publisher_avg = '''
            SELECT AVG(count_reviews)
            FROM (
            SELECT username, COUNT(*) count_reviews
            FROM reviews
            WHERE username IN
            (
            SELECT username--, COUNT(*)
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 50
            )
            GROUP BY username) rev
        

        '''
select(publisher_avg).round()

avg
0  24.0

Пользователи которые оставили более 50 оценок, в среднем написали 24 обзора

# Вывод

После 1 января 2000 г. вышла 821 книга

Посчитали для каждой книги количество обзоров и среднюю оценку

Издательство Penguin Books выпустило больше всего книг

Автор с самой высокой средней оценкой книг J.K. Rowling/Mary GrandPré

Пользователи которые оставили более 50 оценок, в среднем написали 24 обзора